app/controllers/application_controller.rb

```diff
@@ -0,0 +1,54 @@
+class ApplicationController < ActionController::Base
+  protect_from_forgery with: :exception
+
+  before_action :require_login, :except => [:index, :new, :create]
+
+  private
+
+  def sign_in(user)
+    user.regenerate_auth_token
+    cookies[:auth_token] = user.auth_token
+    @current_user = user
+  end
+
+  def permanent_sign_in(user)
+    user.regenerate_auth_token
+    cookies.permanent[:auth_token] = user.auth_token
+    @current_user = user
+  end
+
+  def sign_out
+    @current_user = nil
+    cookies.delete(:auth_token)
+  end
+
+  def current_user
+    @current_user ||= User.find_by_auth_token(cookies[:auth_token]) if cookies[:auth_token]
+  end
+  helper_method :current_user
+
+  def signed_in_user?
+    !!current_user
+  end
+  helper_method :signed_in_user?
+
+  def full_name(user)
+    "#{user.profile.first_name} #{user.profile.last_name}"
+  end
+  helper_method :full_name
+
+  def require_login
+    unless signed_in_user?
+      flash.notice = "Not authorized, please sign in!"
+      redirect_to login_path
+    end
+  end
+
+  def require_current_user
+    unless params[:id] == current_user.id.to_s
+      flash[:error] = "You're not authorized to perform that action."
+      redirect_to user_timeline_path(current_user)
+    end
+  end
```

Comment:
It would be helper to have a method that redirects people when they try to go to a user that doesn't exist (by putting in an ID that doesn't exist) instead of giving them the 404 error page.